packages/userstorage/src/browser/user-storage-frontend-module.ts

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { ContainerModule, } from 'inversify';
+import { ResourceResolver } from '@theia/core/lib/common';
+import { UserStorageResolver } from './user-storage-resource';
+import { UserStorageServiceFilesystemImpl } from './user-storage-service-filesystem';
+import { UserStorageService } from './user-storage-service';
+
+export default new ContainerModule(bind => {
+
+    bind(UserStorageResolver).toSelf().inSingletonScope();
+    bind(ResourceResolver).toDynamicValue(ctx => ctx.container.get(UserStorageResolver));
```

Comment:
`bind(ResourceResolver).to(UserStorageResolver).inSingletonScope();` should do

Reply:
it wont since inversify doesn't support transitive refs his code is ok

Reply:
It would work also there is no other client for UserStorageResolver. Not sure what you mean.

Reply:
To be completely honest I just used the same way ResourceResolver was binded in all the other packages when I prototyped this user storage extension. So I don't know what the best approach is.

Reply:
For example: https://github.com/theia-ide/theia/pull/763#discussion_r150217818
@akosyakov has a ticket about this in inversify see: https://github.com/inversify/InversifyJS/issues/583

Reply:
Yes, but in that case `StatusBarImpl` is injected somewhere. Here we don't inject `UserStorageResolver`, so there is no need to register a binding for that. 
It is important that we understand inversify to that extend and not only copy patterns.
To be clear, we can replace both lines with just the one I suggested in this case. For StatusBarImpl we cannot because it is injected into the shell.

Reply:
Right since it's not injected it's OK. I'm having the hardest time getting this into my head so I'll write it here and in my notes once and for all..

```
bind(UserStorageResolver).toSelf().inSingletonScope();
```

This binds the UserStorageResolver identifier to a singleton using the class of the same name.

```
bind(ResourceResolver).to(UserStorageResolver).inSingletonScope();
```

This binds the ResourceResolver identifier to a singleton (So only 1 RessourceResolver) using the Class UserStorageResolver and not it's identifier.  This is the important part. it's the Class not the identifier.

So in the contribution-provider.ts when doing :

```
if (this.container.isBound(this.serviceIdentifier)) {
                try {
                    this.services = this.container.getAll(this.serviceIdentifier);
```

Here we're doing a get on identifier `ResourceResolver` thus we're getting a instance of `UserStorageResolver` created.

So if we were to inject `UserStorageResolver` somewhere else. We would get 2 instances of `UserStorageResolver`

Actually I wonder if we could/should omit this line:

`bind(UserStorageResolver).toSelf().inSingletonScope()`

Is it doing anything ?

So if we wanted to inject `UserStorageResolver` somewhere else we would have to do:

`bind(ResourceResolver).toDynamicValue(ctx => ctx.container.get(UserStorageResolver))`

SInce this way it will bind the identifier RessourceResolver not the class `UserStorageResolver`  but to the identifier `UserStorageResolver` and since this once is resolved as:

`bind(UserStorageResolver).toSelf().inSingletonScope();`

You will get a singleton `UserStorageResolver` for the contribution in the `getAll` call.

@svenefftinge  Did I get that right ?

Reply:
Yes. And we indeed should remove the line `bind(UserStorageResolver).toSelf().inSingletonScope()`.

Reply:
Ok I think I get it now, thanks for the detailed explanation guys :)